reactive/kotlinx-coroutines-rx2/test/SchedulerTest.kt

```diff
@@ -31,4 +38,252 @@ class SchedulerTest : TestBase() {
+        val disposable = scheduler.scheduleDirect({
+            expectUnreached()
+        }, delayMillis, TimeUnit.MILLISECONDS)
+        delay(100)
+        expect(2)
+        disposable.dispose()
+        delay(300)
+        yield()
+        finish(3)
+    }
+
+    @Test
+    fun testImmediateDispose(): Unit = runTest {
+        expect(1)
+        val scheduler = (currentDispatcher() as CoroutineDispatcher).asScheduler()
+        val disposable = scheduler.scheduleDirect {
+            expectUnreached()
+        }
+        disposable.dispose()
+        yield()
+        finish(2)
+    }
+
+    @Test
+    fun testAsSchedulerWorksWithSchedulerCoroutineDispatcher(): Unit = runTest {
+        expect(1)
+
+        val dispatcher = Schedulers.io().asCoroutineDispatcher()
+        val scheduler = dispatcher.asScheduler()
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect {
+                expect(2)
+                it.resume(Unit)
+            }
+        }
+        finish(3)
+    }
+
+    @Test
+    fun tesConvertDispatcherToOriginalScheduler(): Unit = runTest {
+        expect(1)
+
+        val originalScheduler = Schedulers.io()
+        val dispatcher = originalScheduler.asCoroutineDispatcher()
+        val scheduler = dispatcher.asScheduler()
+        assertEquals(originalScheduler, scheduler)
+
+        finish(2)
+    }
+
+    @Test
+    fun tesConvertSchedulerToOriginalDispatcher(): Unit = runTest {
+        expect(1)
+
+        val originalDispatcher = currentDispatcher() as CoroutineDispatcher
+        val scheduler = originalDispatcher.asScheduler()
+        val dispatcher = scheduler.asCoroutineDispatcher()
+        assertEquals(originalDispatcher, dispatcher)
+
+        finish(2)
+    }
+
+    @Test
+    fun testExpectRxPluginsCall(): Unit = runTest {
+        expect(1)
+
+        fun setScheduler(expectCountOnRun: Int) {
+            RxJavaPlugins.setScheduleHandler(Function {
+                Runnable {
+                    expect(expectCountOnRun)
+                    it.run()
+                }
+            })
+        }
+
+        val dispatcher = currentDispatcher() as CoroutineDispatcher
+        val scheduler = dispatcher.asScheduler()
+
+        setScheduler(2)
+
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect {
+                expect(3)
+                it.resume(Unit)
+            }
+        }
+
+        setScheduler(4)
+        suspendCancellableCoroutine<Unit> {
+            scheduler.scheduleDirect({
+                expect(5)
+                RxJavaPlugins.setScheduleHandler(null)
+                it.resume(Unit)
+            }, 300, TimeUnit.MILLISECONDS)
+        }
+        scheduler.shutdown()
+        finish(6)
+    }
+
+    /**
+     * Let's test the [Scheduler.Worker] to make sure it satisfies the documented constraint of running all work
+     * sequentially
+     */
+    @Test
+    fun testSchedulerWorkerSequentialOrdering(): Unit = runTest {
```

Comment:
Should probably also have a test for the delayed version of this that ensures tasks scheduled at the same time don't overlap. 

Also a test that ensures that delays are actually respected (tasks scheduled sooner in the future run before tasks scheduled later, even when the later task is submitted before the earlier one). Such a test would have caught the misplaced delay call above.